docs/path-resolution.rst

```diff
@@ -0,0 +1,454 @@
+When the source unit name is a relative path, this results in the file being looked up in the
+directory the compiler has been invoked from.
+It is also the only value that results in absolute paths in source unit names being actually
+interpreted as absolute paths on disk.
+
+If the base path itself is relative, it is also interpreted as relative to the current working
+directory of the compiler.
+
+.. index:: ! remapping; import, ! import; remapping, ! remapping; context, ! remapping; prefix, ! remapping; target
+.. _import-remapping:
+
+Import Remapping
+================
+
+Base path and relative imports on their own allow you to freely move your project around the
+filesystem but force you to keep all files within a single directory and its subdirectories.
+When using external libraries it is often desirable to keep their files in a separate location.
+To help with that, the compiler provides another mechanism: import remapping.
+
+Remapping allows you to have the compiler replace import path prefixes with something else.
+For example you can set up a remapping so that everything imported from the virtual directory
+``github.com/ethereum/dapp-bin/library`` would actually receive source unit names starting with
+``dapp-bin/library``.
+By setting base path to ``/project`` you could then have the compiler find them in
+``/project/dapp-bin/library``
+
+The remappings can depend on a context, which allows you to configure packages to import,
+e.g. different versions of a library of the same name.
+
+.. warning::
+
+    Information about used remappings is stored in contract metadata so modifying them will result
+    in a slightly different bytecode.
+    This means that if you move your project files to different locations and use remappings to
+    avoid having to adjust the source code, your project will compile but will no longer produce the
+    exact same bytecode as without the remappings.
+
+Import remappings have the form of ``context:prefix=target``.
+All files in or below the ``context`` directory that import a file that starts with ``prefix`` are
+redirected by replacing ``prefix`` with ``target``.
+For example, if you clone ``github.com/ethereum/dapp-bin/`` locally to ``/project/dapp-bin``,
+you can use the following in your source file:
+
+::
+
+    import "github.com/ethereum/dapp-bin/library/iterable_mapping.sol" as it_mapping;
+
+Then run the compiler:
+
+.. code-block:: bash
+
+    solc github.com/ethereum/dapp-bin/=dapp-bin/ --base-path /project source.sol
+
+As a more complex example, suppose you rely on a module that uses an old version of dapp-bin that
+you checked out to ``/project/dapp-bin_old``, then you can run:
+
+.. code-block:: bash
+
+    solc module1:github.com/ethereum/dapp-bin/=dapp-bin/ \
+         module2:github.com/ethereum/dapp-bin/=dapp-bin_old/ \
+         --base-path /project \
+         source.sol
+
+This means that all imports in ``module2`` point to the old version but imports in ``module1``
```

Comment:
I do not really understand `context` here. I understand the `prefix=target` part, but what does `context` really mean? How can it be used?

Reply:
After reading the rest of this file, I understood how this works. But I didn't understood it when I read about this the first time here. Maybe it makes sense to put here also an example, e.g.:
```
Parts of file and directory names can match as well.
/newProject/con:/new=old will match /newProject/contract.sol 
and remap it to oldProject/contract.sol.
```
But I'm still not sure how this mechanism could be useful.

Reply:
I have rewritten that section and also added this, which hopefully makes the purpose of `context` clearer:

```
You can limit the scope of a remapping by specifying a *context*.
This allows creating remappings that apply only to imports located in a specific library or a specific file.
Without a context a remapping is applied to every matching import in all the files in the virtual
filesystem.
```

> Maybe it makes sense to put here also an example, e.g.:

I don't really want to draw attention to this because for me it's more like an unintended consequence than a feature. If we were to simplify remappings I would actually make them match only the whole segments ending with `/`. I think the current behavior is completely unexpected for most people (it was for me).